### PR TITLE
[Bugfix] Avoid to warmup when world size is 1

### PR DESCRIPTION
**Pull request by kerthcet**
FILL IN THE PR DESCRIPTION HERE

xref: https://github.com/vllm-project/vllm/issues/4587

@youkaichao I have no env to test with the vGPU+vLLM, but at least we can avoid to warmup when the worldsize = 1, do you agree?

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by youkaichao** [comment]
I agree with the idea, but why this PR touches so many lines?

You just need to avoid the warmup if world size > 1.

File: vllm/config.py [hunk 1]
```diff
@@ -29,8 +29,8 @@ class ModelConfig:
 
     Args:
         model: Name or path of the huggingface model to use.
-            It is also used as the content for `model_name` tag in metrics 
-            output when `served_model_name` is not specified.
```
Comment (kerthcet) [inline vllm/config.py:33]: Most of them are format problem, see here. It's auto formatted by my editor.

Comment (kerthcet) [reply vllm/config.py:33]: You have to check the diff from the review mode.

**Comment by kerthcet** [comment]
Ok, got your idea about avoid warmup once worldsize is 1, what I did here is slightly different. Will modify as you suggested.

**Comment by kerthcet** [comment]
Done

**Comment by youkaichao** [comment]
can you revert the format change? let's keep the pr clean on what is changed.

File: vllm/distributed/parallel_state.py [hunk 2]
```diff
@@ -113,8 +113,14 @@ def init_distributed_environment(
                 local_rank = envs.LOCAL_RANK
             else:
                 local_rank = rank
+
         global _LOCAL_RANK
         _LOCAL_RANK = local_rank
+
+        if world_size == 1:
+            # No need to warmup if world size is 1.
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:121]:
```suggestion
            # avoid warmup if world size is 1, to be compatible with vGPU
            # see https://github.com/vllm-project/vllm/issues/4587 for details.
```

**Comment by kerthcet** [comment]
> can you revert the format change? let's keep the pr clean on what is changed.

Make sense to me, I used to make them separate commits. Anyway, reverted.

File: vllm/distributed/parallel_state.py [hunk 3]
```diff
@@ -115,6 +115,11 @@ def init_distributed_environment(
                 local_rank = rank
         global _LOCAL_RANK
         _LOCAL_RANK = local_rank
+
+        if world_size == 1:
+            # Avoid to warmup if world size is 1, to be compatible with vGPU.
```
Comment (youkaichao) [inline vllm/distributed/parallel_state.py:120]:
```suggestion
            # Avoid to warmup if world size is 1, to be compatible with vGPU.
            # see https://github.com/vllm-project/vllm/issues/4587 for details.
```

please leave pointers for future code readers

Comment (kerthcet) [reply vllm/distributed/parallel_state.py:120]: done

**Review by youkaichao — APPROVED** [review]
Thanks for addressing my comments! It can be merged once it passed all the tests.

**Comment by kerthcet** [comment]
Thanks for the quick reply.

**Comment by kerthcet** [comment]
It's weird, the test `buildkite/ci/pr` pending for several hours, still not finished.

**Comment by kerthcet** [comment]
Seems we totally removed the warmup in https://github.com/vllm-project/vllm/pull/5293, can I know the reason here? @youkaichao

**Comment by youkaichao** [comment]
oh, I just happen to remove the warmup because I don't want the warmup for world/tp/pp group. sorry for overwriting your PR!

**Comment by kerthcet** [comment]
Make sense, seems we can close now.

**Pull request closed without merging**